src/tss2-esys/esys_crypto_ossl.c

```diff
@@ -0,0 +1,1114 @@
+    }
+    IESYS_CRYPTOSSL_CONTEXT *mycontext = (IESYS_CRYPTOSSL_CONTEXT *) context;
+    if (mycontext->type != IESYS_CRYPTOSSL_TYPE_HASH) {
+        LOG_ERROR("bad context");
+        return TSS2_ESYS_RC_BAD_REFERENCE;
+    }
+
+    LOGBLOB_TRACE(buffer, size, "Updating hash with");
+
+    if (1 != EVP_DigestUpdate(mycontext->hash.ossl_context, buffer, size)) {
+        return_error(TSS2_ESYS_RC_GENERAL_FAILURE, "OSSL hash update");
+    }
+
+    return TSS2_RC_SUCCESS;
+}
+
+/** Update the digest value of a digest object from a TPM2B object.
+ *
+ * The context of a digest object will be updated according to the hash
+ * algorithm of the context.
+ * @param[in,out] context The context of the digest object which will be updated.
+ * @param[in] b The TPM2B object for the update.
+ * @retval TSS2_RC_SUCCESS on success.
+ * @retval TSS2_ESYS_RC_BAD_REFERENCE for invalid parameters.
+ */
+TSS2_RC
+iesys_cryptossl_hash_update2b(IESYS_CRYPTO_CONTEXT_BLOB * context, TPM2B * b)
+{
+    LOG_TRACE("called for context-pointer %p and 2b-pointer %p", context, b);
+    if (context == NULL || b == NULL) {
+        LOG_ERROR("Null-Pointer passed");
+        return TSS2_ESYS_RC_BAD_REFERENCE;
+    }
+    TSS2_RC ret = iesys_cryptossl_hash_update(context, &b->buffer[0], b->size);
+    return ret;
+}
+
+/** Get the digest value of a digest object and close the context.
+ *
+ * The digest value will written to a passed buffer and the resources of the
+ * digest object are released.
+ * @param[in,out] context The context of the digest object to be released
+ * @param[out] buffer The buffer for the digest value (caller-allocated).
+ * @param[out] size The size of the digest.
+ * @retval TSS2_RC_SUCCESS on success.
+ * @retval TSS2_ESYS_RC_BAD_REFERENCE for invalid parameters.
+ * @retval TSS2_ESYS_RC_GENERAL_FAILURE for errors of the crypto library.
+ */
+TSS2_RC
+iesys_cryptossl_hash_finish(IESYS_CRYPTO_CONTEXT_BLOB ** context,
+                            uint8_t * buffer, size_t * size)
+{
+    LOG_TRACE("called for context-pointer %p, buffer %p and size-pointer %p",
+              context, buffer, size);
+    if (context == NULL || *context == NULL || buffer == NULL || size == NULL) {
+        LOG_ERROR("Null-Pointer passed");
+        return TSS2_ESYS_RC_BAD_REFERENCE;
+    }
+    IESYS_CRYPTOSSL_CONTEXT *mycontext = * context;
+    if (mycontext->type != IESYS_CRYPTOSSL_TYPE_HASH) {
+        LOG_ERROR("bad context");
+        return TSS2_ESYS_RC_BAD_REFERENCE;
+    }
+
+    if (*size < mycontext->hash.hash_len) {
+        LOG_ERROR("Buffer too small");
+        return TSS2_ESYS_RC_BAD_REFERENCE;
+    }
+
+    if (1 != EVP_DigestFinal_ex(mycontext->hash.ossl_context, buffer, NULL)) {
+        LOG_ERROR("Ossl error.");
+        return TSS2_ESYS_RC_GENERAL_FAILURE;
+    }
+
+    LOGBLOB_TRACE(buffer, mycontext->hash.hash_len, "read hash result");
+
+    *size = mycontext->hash.hash_len;
+    // TODO SIZE CHECK memmove(buffer, cpHash, *size);
```

Comment:
EVP_DigestFinal_ex() returns size in the last param if it is not null.
We should just add a local_size variable, pass it to EVP_DigestFinal_ex()
and then assert that local_size == mycontext->hash.hash_len;

Reply:
You are correct. Updated.